htlcswitch/link.go

```diff
@@ -2404,6 +2438,64 @@ func (l *channelLink) handleUpstreamMsg(msg lnwire.Message) {
 		// Update the mailbox's feerate as well.
 		l.mailBox.SetFeeRate(fee)
 
+	case *lnwire.Stfu:
+		handleErr := func(err error) {
+			l.log.Error(err.Error())
+			l.fail(
+				LinkFailureError{
+					code:             ErrInvalidUpdate,
+					FailureAction:    LinkFailureDisconnect,
+					PermanentFailure: false,
+					Warning:          true,
+				},
+				err.Error(),
+			)
+		}
+
+		remoteOnLocal := l.channel.NumRemoteUpdatesPendingOnLocal()
+		remoteOnRemote := l.channel.NumRemoteUpdatesPendingOnRemote()
+		if remoteOnLocal > 0 || remoteOnRemote > 0 {
+			// This is invalid according to the spec, drop the
+			// connection.
+			err := fmt.Errorf(
+				"received stfu while remote updates pending "+
+					"on Local (%d) or Remote (%d)",
+				remoteOnLocal,
+				remoteOnRemote,
+			)
+			handleErr(err)
+
+			return
+		}
+
+		finished, err := l.quiescer.recvStfu()
+		if err != nil {
+			handleErr(err)
+			return
+		}
+
+		if finished {
+			return
```

Comment:
IIUC the goal here is to disconnect the link for a refresh? Simply returning here won't achieve that. However the link can use the new `LinkFailureDisconnect`, type, along with `Warning` false, so it'll function as an internal error, and force the recycle attempt.

Reply:
No we specifically *don't* want to disconnect here. It is at this point that the channel is considered quiescent (frozen) and then we would want to signal to any protocol that is waiting on quiescence that it is allowed to continue. However, no such protocols exist right now so this is empty.